### ISSUE TITLE
Ensure width of table is not wider than console

### ISSUE THREAD

**Issue by tdwright** [tdwright/contabs#6]

**Comment by Chandler-Davidson** [comment]
What's the expected out come when the console window is smaller than the table?

**Comment by GiorgosBetsos** [comment]
Really nice piece of work! How about prompting the user to select either of these options in case the width of the table is bigger than the width of the console:

1. Auto-resize the console window width so that the table fits inside the window
2. Choose one of the available LongStringBehaviour options:
     a. Truncate 
     b. TruncateWithEllipsis 
     c. Wrap
 in order to truncate the table to an appropriate width.

**Comment by tdwright** [comment]
Thanks @GiorgosBetsos - glad you like it!

I'm not sure whether your first option would work in all cases, but this is just a gut feeling.

Your second option(s) seems to me to have more potential. Do you envisage these methods being used on each line of the table, or on values in individual cells? If the latter, how would we identify which cells to operate on?

As I've previously commented, I don't have a clear picture in my head about how this _should_ work. I've just seen it go wrong so many times that I'm keenly aware of the need to do something.

If you're keen to take this on, I might suggest going low tech and experimenting in notepad. See what approach you like and then see if you can persuade me of its merits! :wink:

**Comment by zumo-sk** [comment]
I think I could give this a try. Or, I have to admit, I already started tackling it. :smile: How I would like to approach the problem:
1. Have a new table property CanvasWidth. Default value is width of console, but user can change it (depending where they want to output it).
2. Have a table long string setting, that can be overwritten for each column.
3. Table will have property TableStretchStyle, that will be used for both stretching and squeezing the table (stretching and squeezing is very similar). I am currently considering these values (order is just for better understanding):
- DoNothing = considering so far as a default behavior, the one we're using now
- AllColumnsEqualWidth = all columns will have approximately the same width to fill up the canvas width
- StretchOrSqueezeAllColumnsEvenly = calculate the optimal width first, then stretch or squeeze all of them evenly
- SqueezeAllColumnsEvenly = the same as previous one, but does not fill the canvas if table is not wider than canvas
- SqueezeLongStrings = will try to squeeze the long string columns first, We have to decide when to start squeezing others (maybe when Column name would not be visible anymore?).
- StretchOrSqueezeLongStrings = Same as above, but stretches long strings if canvas is wider than table. Stretching ratio could be the same as your original long string column width ratios.

And one more bonus that is easy to make - when we have CanvasWidth, we can actually align table itself (to be centered or aligned right), if we decide not to stretch it.

**Comment by tdwright** [comment]
Wow @zumo-sk - you've really given this some thought!

_Sorry about the delayed response. I was off grid in deepest, darkest Tasmania..._

One thing that strikes me immediately is that we don't currently have a way of handling very wide non-string columns. I'm not near an IDE to test this, but I think that `LongStringBehaviour` is ignored for non-string columns?

Another thing that springs to mind is that we may not need to explicitly address _column_ width in order to deliver some value on this issue. The minimum we'd need to do is to prevent the wrapping behaviour we're observing. The simplest way to do this (and a good starting point) would be to simply truncate the table.

The default behaviour of all columns is already to take up the least width needed, after taking formatting into account. Perhaps this means that our strategies for _fitting_ the table into the space should focus on the way we handle long strings. In fact, it seems as though this has already been your starting point?

Tom

**Comment by zumo-sk** [comment]
Truncating the table is an excellent idea and I will definitely use it as a default behavior. I don't know why I haven't realized truncating the table would be an option, too - it's always better than breaking the table.

I think I'll do it in phases - the first phase would be to truncate the table if necessary and to implement the behaviors for stretching / squeezing the table. I progressed significantly already, but I have holiday season now and I'm barely near computer outside work, so it will take some time. Sorry. But I'm on it! :)

Second phase then will be adapting long strings for other types than strings (it seemed super easy to do) and add some other long string behaviors, like HTML table style word-breaks. But first things first :)

**Comment by tdwright** [comment]
This all sounds great @zumo-sk. Perhaps break out the phases into a checklist? Or, if you'd prefer, create new issues for future work.

I'm thinking about whether to include this in an upcoming milestone: [v2.0](https://github.com/tdwright/contabs/milestone/1). What do you think? Should we aim to include some, or all, of this functionality? It mostly depends on you capacity, so let me know what sort of timeframe you think you'd need to do what you want to do, and I'll plan accordingly.

**Comment by zumo-sk** [comment]
I still consider myself a developer, so I am terrible at evaluations. But let's see. 😄
It depends on when do you want to release v2.0 - if this year, there is no way of finalizing it along with tests due to vacations and holidays. My target is Jan 2019.
Things I really want to implement as part of this ticket:
- [ ] Prevent columns that would be outside of console to be drawn
- [ ] Implement stretch / squeeze algorithms, truncating the non-long string texts
- [ ] Implement table alignment - center the whole table (not columns), or align to the right
I have other plans, but I don't want to aim too high. Sounds good?

**Comment by zumo-sk** [comment]
Hi Tom,

I have implemented a solution, but I found out I have some technical issues caused by lack of previous open source participation.
1. Do you have an advice how to check unit test coverage without having VS Enterprise? I was very disappointed to see it missing in professional edition. I don't want to install a lot of extensions before actually knowing whether it's worth it.
2. Is it common to offer something for design review or "please review this before I write tests" review? Or to discuss some decisions found during implementation? If so, is the best way to do it raising a pull request?
3. I found some trickier places to be commented in the code (eg. long string word wrap methods). Do you have a rule of thumb which things have to have comments and which ones will survive without them?

Thank you in advance, both for your answer and patience. I'm going to visit the Big Apple in following days, so I might not be very responsive.

Marek

**Comment by tdwright** [comment]
Hi Marek,

That was quick! Wasn't expecting anything til Jan.

The answer to your second question is a "yes". In fact, this will help with your other questions too.

Code coverage is a pain in the ass to set up without paying for something. I don't really expect anyone to do code coverage analysis before opening a PR, as long as I can see they've made a good-faith effort to ensure all their new code is tested.

If you really want to have a go at it on your local machine (hey, I commend your curiosity!) then a good place to start would be the [AppVeyor config file](https://github.com/tdwright/contabs/blob/master/appveyor.yml#L17). This is what drives the automated analysis as part of the CI system and only uses free tools: namely NUnit (with which you're presumably already familiar) and OpenCover. You should be able to run those same commands from a terminal on your own machine and be able to generate a coverage.xml file. To understand this file, I'd recommend [ReportGenerator](https://github.com/danielpalme/ReportGenerator), which turns the XML into a bundle of webpages.

As for comments... I try to be practical. My natural inclination is to comment perhaps more than some would recommend, but I'd rather code was readable. The downside of comments is that they sometimes become out of date, which can make tricky code even harder to understand. I think perhaps it would be useful to see the comments your talking about so I can get a better idea of the sort of thing you're talking about.

Hope that help!

Tom

**Comment by tdwright** [comment]
Oh, and have a great time in NY! 😃

**Comment by zumo-sk** [comment]
Thank you for a quick reply!
Yes, I had a free Saturday, so I spent quality time with some development. 😄 I found out some of my assumptions were not exactly correct (for example, now I think that you don't want to have non-long string columns squeezed unless you specifically ask for it) and I developed things slightly differently than I originally planned, but I'll write about it once I write tests and create a merge request.
OK, I decided not to try to run the code coverage locally.. yet. 😄 I'll write just tests that I think make sense and to have a good gut feeling and I'll make a pull request - and we can take it from there.

And thanks, I will definitely enjoy NY!
